### ISSUE TITLE
Analyzer doesn't report "Generic function type cannot be used as a type parameter"

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#33555]
Consider the following code
```dart
dynamic forgetType(dynamic d) => d;

typedef T0 = void Function();
typedef T1 = void Function<X>();

void t0Instance() {}
void t1Instance<X>() {}

class ArgumentsBindingTest<X>  {
  ArgumentsBindingTest(X t1) {}
}

main() {
  T0 t0 = t0Instance;
  T1 t1 = t1Instance;
  new ArgumentsBindingTest<T0>(forgetType(t0)); // works
  new ArgumentsBindingTest<T1>(forgetType(t1)); // error
}
```
This code produces
```
error: generic function type '<X>() => void' not allowed as type argument of type 'ArgumentsBindingTest<<X>() => void>'
#0      _startIsolate.<anonymous closure> (dart:isolate/runtime/libisolate_patch.dart:279:19)
#1      _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:165:12)
```

Why function type can be used as a type parameter, but generic function type cannot?

Dart VM version: 2.0.0-dev.63.0 (Fri Jun 15 00:42:43 2018 +0200) on "windows_x64"

**Comment by lrhn** [comment]
The class `Exception` doesn't cover most errors in Dart. You probably want to use `Error` instead (or just drop the `on` part of the catch and catch all thrown objects).

This does throw a `TypeError`, which implements `Error` and not `Exception`. I don't know why it prints badly, if you paste the code into [DartPad](http://dartpad.dartlang.org/), it prints the error message as
```
TypeError: Closure 't0Func': type '<T1 extends B0, T2 extends B1>(V0, V1, {V2 x2, V3 x3, V4 x4}) => U0' is not a subtype of type '<T1 extends B0, T2 extends B1>(S0, S1, {S2 x2, S3 x3, S4 x4, V4 x5}) => U1'
```

**Comment by sgrekhov** [comment]
@lrhn, problem is that  generic function type cannot be used as a type parameter. I've rewritten the issue. Please, look at it once again

**Comment by lrhn** [comment]
That's actually expected behavior. We do not allow generic method types as type arguments.

**Comment by sgrekhov** [comment]
If this is an expected behaviour, then, there should be an error in analyzer, I think, but there isn't.

**Comment by lrhn** [comment]
There should be an error reported by analyzer, yes.

**Comment by scheglov** [comment]
I believe it was fixed in 36400371646b5a3290d4dc3d1fa8ea81b28fe997 and 284b5cce55f96dad08d3402dacae5e81f1e9351a.
